File: C++/200_Number_of_Islands.cpp
class Solution {
    void bfs(vector<vector<char>>& grid,  map<pair<int, int> , bool>& visited, pair<int, int> s){
        queue<pair<int, int>> q;
        q.push(s);
        vector<pair<int, int>> v {{0, -1}, {-1, 0}, {0, 1}, {1, 0}};
        while(!q.empty()){
            pair<int, int> u = q.front();
            q.pop();
            for(pair<int, int> x : v){
                int nx = u.first + x.first;
                int ny = u.second + x.second;
                if(nx >= 0 && nx <= grid.size() - 1 && ny >= 0 && ny <= grid[0].size() - 1){
                    if(grid[nx][ny] == '1' && visited.count({nx, ny}) == 0){
                        visited[{nx, ny}] = 1;
                        q.push({nx, ny});
                    }
                }
            }
        }
    }
    
public:
    int numIslands(vector<vector<char>>& grid) {
        map<pair<int, int> , bool> visited;
        int ans = 0;
        for(int i = 0; i < grid.size(); i++){
            for(int j = 0; j < grid[0].size(); j++){
                if(grid[i][j] == '1' && visited.count({i, j}) == 0){
                    visited[{i, j}] = 1;
                    bfs(grid, visited, {i, j});
                    ans += 1;
                }
            }
        }
        return ans;
    }
};

File: Python/200_Number_of_Islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        k = 0
        visited = {}
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if (i, j) not in visited and grid[i][j] == '1':
                    self.BFS(grid, (i, j), visited)
                    k += 1            
        return k
    
    def BFS(self, G, s, visited):  
        queue = collections.deque()
        queue.append(s)
        visited[s] = True
        while queue:
            u = queue.popleft()
            for v in ((u[0], u[1]+1), (u[0]+1, u[1]), (u[0], u[1]-1), (u[0]-1, u[1])):
                if v not in visited and v[0] < len(G)  and v[0] >= 0 and v[1] < len(G[0]) and v[1] >= 0:
                    if G[v[0]][v[1]] == '1':
                        visited[v] = True
                        queue.append(v)
